onnxruntime/core/providers/cann/gpu_data_transfer.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/shared_library/provider_api.h"
+#include "core/providers/cann/gpu_data_transfer.h"
+#include "core/providers/cann/cann_call.h"
+
+namespace onnxruntime {
+GPUDataTransfer::GPUDataTransfer(aclrtStream stream, bool do_copy_in_default_stream) {
+  // create streams, default is nullptr
+  do_copy_in_default_stream_ = do_copy_in_default_stream;
+  streams_[kCannStreamDefault] = stream;
+  if (do_copy_in_default_stream) {
+    streams_[kCannStreamCopyIn] = stream;
+    streams_[kCannStreamCopyOut] = stream;
+  } else {
+    CANN_CALL_THROW(aclrtCreateStream(&streams_[kCannStreamCopyIn]));
+    CANN_CALL_THROW(aclrtCreateStream(&streams_[kCannStreamCopyOut]));
+  }
+}
+
+GPUDataTransfer::~GPUDataTransfer() {
+  if (!do_copy_in_default_stream_ && streams_[kCannStreamCopyIn] != nullptr) {
+    CANN_CALL_THROW(aclrtDestroyStream(streams_[kCannStreamCopyIn]));
+  }
+  if (!do_copy_in_default_stream_ && streams_[kCannStreamCopyOut] != nullptr) {
+    CANN_CALL_THROW(aclrtDestroyStream(streams_[kCannStreamCopyOut]));
+  }
+}
+
+bool GPUDataTransfer::CanCopy(const OrtDevice& src_device, const OrtDevice& dst_device) const {
+  return src_device.Type() == OrtDevice::GPU || src_device.MemType() == OrtDevice::MemType::CANN_PINNED ||
+         dst_device.Type() == OrtDevice::GPU || dst_device.MemType() == OrtDevice::MemType::CANN_PINNED;
+}
+
+common::Status GPUDataTransfer::CopyTensor(const Tensor& src, Tensor& dst, int exec_queue_id) const {
+  size_t bytes = src.SizeInBytes();
+  const void* src_data = src.DataRaw();
+  void* dst_data = dst.MutableDataRaw();
+
+  auto& src_device = src.Location().device;
+  auto& dst_device = dst.Location().device;
+
+  if (dst_device.Type() == OrtDevice::GPU) {
```

Comment:
are we using a tool similar to AMD's HIP to convert the ort cuda EP to this EP? THe code here looks exactly the same as cuda EP.

Reply:
oh, thanks for point out this. Ascend is not GPU, it's a NPU device. We reused GPU type before which is not correct. There are three device type in onnxruntime (CPU GPU FPGA), how about add a new one called NPU? It's good for other NPU integration as well.